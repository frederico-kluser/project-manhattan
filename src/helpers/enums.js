var enums = {
  colors: {
    black: '#212121',
    white: '#FFFFFF',
  },
  helpers: {
    infoClass: 'kluser_element-info noselect',
  },
  icons: {
    add_circle: 'add_circle',
    code: 'code',
    keyboard: 'keyboard',
    search: 'search',
  },
  mod: {
    sizeMode: 'sizeMode',
    moveMode: 'moveMode',
  },
};

window.enums = enums;
