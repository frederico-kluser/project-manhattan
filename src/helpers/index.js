/* eslint-disable-next-line complexity */
const createElementor = ({attributes = [], chield = [], className, html, id, tag, text, style}) => {
  const element = document.createElement(tag);
  if (className) {
    element.setAttribute('class', className);
  }
  if (attributes.length) {
    attributes.forEach(({attribute, value}) => {
      element.setAttribute(attribute, value);
    });
  }
  if (chield.length) {
    chield.forEach(item => {
      element.appendChild(item);
    });
  }
  if (text) {
    element.innerText = text;
  }
  if (id) {
    element.setAttribute('id', id);
  }
  if (style) {
    element.setAttribute('style', style);
  }
  if (html) {
    element.innerHTML = html;
  }

  return element;
};

const dynamicFunction = (func1, func2, conditional) => {
  let func = func1 || func2;

  if (conditional === false) {
    func = func2;
  }

  return func();
};

const backSpaceText = string => string.substring(0, string.length - 1);

window.createElementor = createElementor;
window.dynamicFunction = dynamicFunction;
window.backSpaceText = backSpaceText;
