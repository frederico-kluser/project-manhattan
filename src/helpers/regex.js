const regex = {
  only_numbers: /^(-?)[0-9]*$/,

  test: (value, rule) => !!rule.exec(value),
  adjut: (value, rule, extension = '') => (rule.exec(value) !== null ? value + extension : value),
};

export default regex;
