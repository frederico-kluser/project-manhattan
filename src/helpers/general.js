/* eslint-disable complexity */

import {enums} from './enums.js';

export const getClassUnicName = className => {
  const unicNameLocation = className.indexOf('kluser_');
  if (unicNameLocation === -1) {
    throw new Error("Don't exist unic kluser_ class");
  }
  return className.substring(unicNameLocation, unicNameLocation + 16);
};

export const unicGlobalVarNameGenerator = () => {
  let name;

  do {
    name = `kluser_${Math.random().toString(36).substr(2, 9)}`;
  } while (document.getElementsByClassName(name).length);

  return name;
};

const getRandomInteger = (min, max) => Math.floor(Math.random() * (max - min + 1) + min);

export const getRandomColor = (intensity = 500) => {
  const colorKey = getRandomInteger(1, Object.keys(enums.colors).length - 3);
  const colorProperty = Object.keys(enums.colors)[colorKey];
  return enums.colors[colorProperty][intensity];
};

export const createElement = ({
  attributes = [],
  chield = [],
  className,
  html,
  id,
  tag,
  text,
  style,
}) => {
  const elem = document.createElement(tag);
  if (className) {
    elem.setAttribute('class', className);
  }
  if (attributes.length) {
    attributes.forEach(({attribute, value}) => {
      elem.setAttribute(attribute, value);
    });
  }
  if (chield.length) {
    chield.forEach(item => {
      elem.appendChild(item);
    });
  }
  if (text) {
    elem.innerText = text;
  }
  if (id) {
    elem.setAttribute('id', id);
  }
  if (style) {
    elem.setAttribute('style', style);
  }
  if (html) {
    elem.innerHTML = html;
  }

  return elem;
};

export const conditionalFunctionExecute = (func1, func2) => {
  let result;

  if (func1) {
    result = func1();
  }

  if (func2 && !result) {
    result = func2();
  }

  return !!result;
};

export const builderMenu = e => {
  e.preventDefault();
};
