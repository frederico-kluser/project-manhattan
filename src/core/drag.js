import {elementIdSetter, Elements} from './elements.js';

export const dragStartEventSetter = e => {
  if (e.which === 1) {
    const {id} = e.target;
    elementIdSetter(id);
    Elements[id].dragEvents(e, 'start');
  }
};

export const dragMoveEventSetter = e => {
  const {id} = e.target;
  Elements[id].dragEvents(e, 'move');
};

export const dragEndEventSetter = e => {
  const {id} = e.target;
  Elements[id].dragEvents(e, 'end');
};
