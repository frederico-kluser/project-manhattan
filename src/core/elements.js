import {enums} from '../helpers/enums.js';
import {
  builderMenu,
  createElement,
  getRandomColor,
  unicGlobalVarNameGenerator,
} from '../helpers/general.js';
import regex from '../helpers/regex.js';
import {propertySizeInPixels} from '../styles/cssProperties.js';
import {
  makeOnlyElementtouchable,
  styleTagUpdater,
  updateHelperBubble,
  fixPropertySize,
} from '../styles/styleGeneral.js';
import {dragEndEventSetter, dragMoveEventSetter, dragStartEventSetter} from './drag.js';

export const Elements = {};

const stylePropertyDefault = () => ({
  border: '1px solid black',
  height: 100,
  left: 100,
  position: 'relative',
  top: 100,
  width: 100,
});

let elementId;
export const elementIdGetter = () => elementId;
export const elementIdSetter = value => {
  elementId = value;
};

export const ElementConfig = (id, element) => {
  const {sizeMode} = enums.mod;
  const {backgroundColor} = enums.style;

  elementIdSetter(id);
  let _mode = sizeMode;
  const _style = stylePropertyDefault();
  _style[backgroundColor] = getRandomColor(500);

  let _onDraggingElement = false;

  let _initialPositionX = 0;
  let _initialPositionY = 0;

  let _positionX = 0;
  let _positionY = 0;

  let _calcValueX = 0;
  let _calcValueY = 0;

  const _element = element;

  const _styleDynamic = {};
  Object.keys(propertySizeInPixels).forEach(property => {
    _styleDynamic[`${property}Absolute`] = _element[propertySizeInPixels[property]];
    _styleDynamic[`${property}Relative`] = _style[property] || 0;
  });

  const _sizeConverter = (type, plus, resumeOutput = false) => {
    // eslint-disable-next-line no-restricted-globals
    const sizeType = isNaN(_styleDynamic[`${type}Relative`])
      ? regex.getSizeType(_styleDynamic[`${type}Relative`])
      : false;
    if (sizeType !== false) {
      const valueType = parseFloat(_styleDynamic[`${type}Relative`]);

      const absoluteSizeBefore = _styleDynamic[`${type}Absolute`];
      const absoluteSizeAfter = absoluteSizeBefore + plus;

      const relativeSizeAfter = (absoluteSizeAfter * valueType) / absoluteSizeBefore;

      return (resumeOutput ? relativeSizeAfter.toFixed(2) : relativeSizeAfter) + sizeType;
    }
    return _style[type] + (plus || '');
  };

  const _helperBubbleTextRender = () => {
    const {moveMode} = enums.mod;
    let output;
    const resumeOutput = true;

    switch (_mode) {
      case sizeMode:
      default:
        output =
          fixPropertySize(_sizeConverter('height', _calcValueY, resumeOutput), 'height', true) +
          fixPropertySize(_sizeConverter('width', _calcValueX, resumeOutput), 'width', true);
        break;
      case moveMode:
        output =
          fixPropertySize(_sizeConverter('top', _calcValueY, resumeOutput), 'top', true) +
          fixPropertySize(_sizeConverter('left', _calcValueX, resumeOutput), 'left', true);
        break;
    }

    return output;
  };

  const _dynamicStylePropertiesTextSetter = () => {
    const {tune, search} = enums.icons;
    let info = _helperBubbleTextRender();
    let icon = tune;

    if (!_onDraggingElement) {
      info = '';
      icon = search;
    }

    updateHelperBubble(info, icon);
  };

  const _fixSizeTypes = property => {
    let output;

    if (regex.test(_style[property], regex.onlyNumbers)) {
      output = parseFloat(_style[property]);
    } else {
      output = _style[property];
    }

    _style[property] = output;
  };

  const _updateDynamicProperties = () => {
    const styleAttributeHTML = _helperBubbleTextRender();

    _fixSizeTypes('height');
    _fixSizeTypes('width');
    _fixSizeTypes('top');
    _fixSizeTypes('left');

    _element.setAttribute('style', styleAttributeHTML);
  };

  const _updateElementStyles = () => {
    const {moveMode} = enums.mod;

    switch (_mode) {
      case sizeMode:
      default:
        _style.width = _sizeConverter('width', _calcValueX);
        _style.height = _sizeConverter('height', _calcValueY);
        break;
      case moveMode:
        _style.left = _sizeConverter('left', _calcValueX);
        _style.top = _sizeConverter('top', _calcValueY);
        break;
    }
  };

  return {
    elementGetter: () => _element,
    dragEvents: (e, type) => {
      const x = e.clientX;
      const y = e.clientY;

      switch (type) {
        case 'start':
          _onDraggingElement = true;

          _initialPositionX = x;
          _initialPositionY = y;

          _calcValueX = 0;
          _calcValueY = 0;

          Object.keys(propertySizeInPixels).forEach(property => {
            _styleDynamic[`${property}Absolute`] = _element[propertySizeInPixels[property]];
            _styleDynamic[`${property}Relative`] = _style[property] || 0;
          });

          styleTagUpdater();
          makeOnlyElementtouchable(_element);
          break;
        case 'move':
          if (_onDraggingElement) {
            _positionX = x;
            _positionY = y;

            _calcValueX = _positionX - _initialPositionX;
            _calcValueY = _positionY - _initialPositionY;

            _updateDynamicProperties();
          }
          break;
        case 'end':
          _onDraggingElement = false;

          _updateElementStyles();

          _calcValueX = 0;
          _calcValueY = 0;

          styleTagUpdater();
          break;
      }

      _dynamicStylePropertiesTextSetter();
    },
    modeSetter: mode => {
      _mode = mode;
    },
    styleGetter: () => _style,
    styleSetter: (command, value) => {
      _style[command] = value;
    },
  };
};

export const ElementBuilder = (parent, tag, className = '', text = '') => {
  const id = unicGlobalVarNameGenerator();
  const element = createElement({className, id, tag, text});
  element.addEventListener('mousedown', dragStartEventSetter);
  element.addEventListener('mousemove', dragMoveEventSetter);
  element.addEventListener('mouseup', dragEndEventSetter);
  element.addEventListener('contextmenu', builderMenu);
  parent.appendChild(element);

  Elements[id] = ElementConfig(id, element);
};
