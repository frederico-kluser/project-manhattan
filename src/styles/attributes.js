import {shortCutCommandsSetter} from '../core/commands.js';
import {elementIdGetter, Elements} from '../core/elements.js';

const setStyle = ({command, value}) => {
  Elements[elementIdGetter()].style[command] = value;
  return true;
};

const attributeBuilder = (
  attribute,
  shiftCommand = false,
  hasSizeType = false,
  sizeInPixels = false
) => {
  const _attributeShiftCommands =
    shiftCommand === true
      ? {[attribute.charAt(0)]: () => shortCutCommandsSetter(attribute)}
      : shiftCommand;
  const _attributesHasSizeType = hasSizeType === true ? {[attribute]: hasSizeType} : hasSizeType;
  const _attributeSizeInPixels =
    sizeInPixels !== false ? {[attribute]: sizeInPixels} : sizeInPixels;

  return {
    [attribute]: {
      _attributeShiftCommands,
      _attributesHasSizeType,
      _attributeCommands: {[attribute]: setStyle},
      _attributeSizeInPixels,
    },
  };
};

const attributeFunctions = {
  ...attributeBuilder('height', true, true, 'clientHeight'),
  ...attributeBuilder('width', true, true, 'clientWidth'),
  ...attributeBuilder('left', true, true, 'offsetLeft'),
  ...attributeBuilder('top', true, true, 'offsetTop'),
  ...attributeBuilder('margin', true, true),
  ...attributeBuilder('padding', true, true),
};

export const attributesWithDynamicSizeArr = ['height', 'width', 'left', 'top', 'margin', 'padding'];
export let attributeShiftCommands = {};
export let attributesHasSizeType = {};
export let attributeCommands = {};
export let attributeSizeInPixels = {};

(() => {
  Object.values(attributeFunctions).forEach(
    ({
      _attributesHasSizeType,
      _attributeCommands,
      _attributeShiftCommands,
      _attributeSizeInPixels,
    }) => {
      if (_attributeShiftCommands) {
        attributeShiftCommands = {
          ..._attributeShiftCommands,
          ...attributeShiftCommands,
        };
      }

      if (_attributesHasSizeType) {
        attributesHasSizeType = {
          ..._attributesHasSizeType,
          ...attributesHasSizeType,
        };
      }

      if (_attributeCommands) {
        attributeCommands = {
          ..._attributeCommands,
          ...attributeCommands,
        };
      }

      if (_attributeSizeInPixels) {
        attributeSizeInPixels = {
          ..._attributeSizeInPixels,
          ...attributeSizeInPixels,
        };
      }
    }
  );
})();
