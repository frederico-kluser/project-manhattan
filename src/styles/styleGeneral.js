import {enums} from '../helpers/enums.js';
import {Elements} from '../core/elements.js';
import regex from '../helpers/regex.js';
import {createElement} from '../helpers/general.js';
import {propertyHasSizeType} from './cssProperties.js';
import {stopKeyboardCommandsSetter} from '../core/commands.js';
import {globalStyleObject, inputExtraSize} from './styleEnums.js';

const {helperBubble, helperInput, helperText} = enums.helperBubble;

const autoReizeInput = ({target}) => {
  const {length} = target.value;
  target.setAttribute('size', length + inputExtraSize);
  target.setAttribute('maxlength', length + inputExtraSize);
};

let styleTagElement;
export const styleTagElementGetter = () => styleTagElement;
export const styleTagElementSetter = value => {
  styleTagElement = value;
};

export const helperBubbleStructure = {};
helperBubbleStructure.symbol = enums.icons.search;
helperBubbleStructure.icon = createElement({
  tag: 'span',
  className: 'material-icons',
  text: helperBubbleStructure.symbol,
});
helperBubbleStructure.text = createElement({tag: 'div', className: helperText});
helperBubbleStructure.input = createElement({tag: 'input', className: helperInput});
helperBubbleStructure.element = createElement({
  tag: 'div',
  className: helperBubble,
  chield: [helperBubbleStructure.icon, helperBubbleStructure.text, helperBubbleStructure.input],
});

export const updateHelperBubble = (
  info = '',
  icon = helperBubbleStructure.symbol,
  inputPlaceholder = '',
  inputType = 'text',
  inputValue = ''
) => {
  const size = inputValue.length + inputPlaceholder.length;
  stopKeyboardCommandsSetter(info);
  helperBubbleStructure.text.innerText = info;
  helperBubbleStructure.symbol = icon;
  helperBubbleStructure.icon.innerText = icon;
  helperBubbleStructure.input.setAttribute('placeholder', inputPlaceholder);
  helperBubbleStructure.input.setAttribute('type', inputType);
  helperBubbleStructure.input.setAttribute('size', size);
  helperBubbleStructure.input.setAttribute('maxlength', size);
  helperBubbleStructure.input.value = inputValue;
  helperBubbleStructure.input.setAttribute('style', inputPlaceholder ? `display: initial;` : '');
  if (inputPlaceholder) {
    // refactor this
    setTimeout(() => {
      helperBubbleStructure.input.focus();
    }, 1);
  }
  helperBubbleStructure.input.addEventListener('keydown', autoReizeInput);
  helperBubbleStructure.element.setAttribute(
    'style',
    info !== '' ? `opacity:1;max-width:500px;border:none;` : 'transition: opacity 0.25s;'
  );
};

export const fixPropertySize = (value, cssAttribute, printMode = false) => {
  const defaultSizeType = 'px';
  const output = propertyHasSizeType[cssAttribute]
    ? regex.insertSizeType(value, regex.onlyNumbers, defaultSizeType)
    : value;

  return printMode ? `${cssAttribute}: ${output};` : output;
};

export const styleTagUpdater = () => {
  let css = Object.values(globalStyleObject).join('');

  Object.keys(Elements).forEach(elementKey => {
    css += `\n.${elementKey} {\n`;
    const element = Elements[elementKey].elementGetter();
    const style = Elements[elementKey].styleGetter();

    element.removeAttribute('style');
    const sortAttributes = [];

    Object.keys(style).forEach(cssAttribute => {
      sortAttributes.push(`\t${fixPropertySize(style[cssAttribute], cssAttribute, true)}\n`);
    });

    sortAttributes.sort();
    sortAttributes.forEach(attribute => {
      css += attribute;
    });

    css += '}\n';
  });

  styleTagElementGetter().innerHTML = css;
  updateHelperBubble();
};

export const injectedCssUpdater = () => {
  styleTagElementSetter(
    createElement({
      tag: 'style',
      attributes: [{attribute: 'type', value: 'text/css'}],
    })
  );
  const linkIcons = createElement({
    tag: 'link',
    attributes: [
      {
        attribute: 'href',
        value: 'https://fonts.googleapis.com/icon?family=Material+Icons',
      },
      {
        attribute: 'rel',
        value: 'stylesheet',
      },
    ],
  });
  const linkFonts = createElement({
    tag: 'link',
    attributes: [
      {
        attribute: 'href',
        value:
          'https://fonts.googleapis.com/css2?family=Roboto+Mono:ital,wght@0,100;0,200;0,300;0,400;0,500;0,600;0,700;1,100;1,200;1,300;1,400;1,500;1,600;1,700&display=swap',
      },
      {
        attribute: 'rel',
        value: 'stylesheet',
      },
    ],
  });

  document.getElementsByTagName('head')[0].appendChild(linkIcons);
  document.getElementsByTagName('head')[0].appendChild(linkFonts);
  document.getElementsByTagName('head')[0].appendChild(styleTagElementGetter());
  document.body.appendChild(helperBubbleStructure.element);
  styleTagUpdater();
};

export const makeOnlyElementtouchable = draggableElement => {
  Object.keys(Elements).forEach(elementKey => {
    const element = Elements[elementKey].elementGetter();
    if (element !== draggableElement && element.contains(draggableElement) === false) {
      element.setAttribute('style', 'pointer-events: none !important;');
    }
  });
};

// eslint-disable-next-line no-unused-vars
const elementCssSelectors = element => {
  const sheets = document.styleSheets;
  const apliedCSS = [];

  // eslint-disable-next-line no-param-reassign
  element.matches =
    element.matches ||
    element.webkitMatchesSelector ||
    element.mozMatchesSelector ||
    element.msMatchesSelector ||
    element.oMatchesSelector;
  Object.values(sheets).forEach(sheet => {
    const rules = sheet.rules || sheet.cssRules;
    Object.values(rules).forEach(rule => {
      if (element.matches(rule.selectorText)) {
        apliedCSS.push(rule.cssText);
      }
    });
  });

  return apliedCSS;
};

// eslint-disable-next-line complexity
export const getAllStyles = element => {
  if (!element) return {}; // Element does not exist, empty list.
  const win = document.defaultView || window;
  const styleNode = {};
  const style = win.getComputedStyle(element, '');
  Object.values(style).forEach(property => {
    styleNode[property] = style.getPropertyValue(property);
  });

  // https://stackoverflow.com/questions/2952667/find-all-css-rules-that-apply-to-an-element
  return styleNode;
};
