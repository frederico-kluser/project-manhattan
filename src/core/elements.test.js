/* eslint-disable no-new */
import {injectedCssUpdater} from '../styles/styleGeneral';
import {ElementBuilder, Elements, elementIdGetter, ElementConfig} from './elements';
import * as general from '../helpers/general';
import * as drag from './drag';

jest
  .spyOn(general, 'unicGlobalVarNameGenerator')
  .mockImplementation()
  .mockReturnValueOnce('mockId1')
  .mockReturnValueOnce('mockId2')
  .mockReturnValueOnce('mockId3');
jest.spyOn(general, 'getRandomColor').mockImplementation().mockReturnValue('#FFFFFF');

describe('function ElementBuilder', () => {
  it('object Elements start a empty object', () => {
    expect.assertions(3);

    expect(Elements).toStrictEqual({});
    expect(Object.keys(Elements)).toHaveLength(0);
    ElementBuilder(document.body, 'div');
    expect(Object.keys(Elements)).toHaveLength(1);
  });

  it('inject children on Elements', () => {
    expect.assertions(4);

    expect(Object.keys(Elements)).toHaveLength(1);
    const elementFather = Elements[elementIdGetter()].elementGetter();

    ElementBuilder(elementFather, 'div');
    const elementSon1 = Elements[elementIdGetter()].elementGetter();

    ElementBuilder(elementFather, 'div');
    const elementSon2 = Elements[elementIdGetter()].elementGetter();

    const quantOfElements = Object.keys(Elements).length;

    expect(quantOfElements).toStrictEqual(3);
    expect(elementFather.childNodes[0]).toStrictEqual(elementSon1);
    expect(elementFather.childNodes[1]).toStrictEqual(elementSon2);
  });

  it('function injectedCssUpdater insert style tag on html head', () => {
    expect.assertions(3);

    const styleTagContent = () =>
      document.getElementsByTagName('head')[0].getElementsByTagName('style')[0];

    expect(styleTagContent()).toBeUndefined();
    injectedCssUpdater();
    expect(styleTagContent()).not.toBeUndefined();
    expect(styleTagContent()).toMatchSnapshot();
  });

  it('take a snapshot of Elements', () => {
    expect.assertions(1);

    expect(Elements).toMatchSnapshot();
  });

  it('test created elements', () => {
    expect.assertions(4);

    const element = Elements[elementIdGetter()].elementGetter();

    const dragStartEventSetterSpied = jest.spyOn(drag, 'dragStartEventSetter');
    const dragMoveEventSetterSpied = jest.spyOn(drag, 'dragMoveEventSetter');
    const dragEndEventSetterSpied = jest.spyOn(drag, 'dragEndEventSetter');

    element.dispatchEvent(new Event('mousedown'));
    element.dispatchEvent(new Event('mousemove'));
    element.dispatchEvent(new Event('mouseup'));

    expect(element).toMatchSnapshot();
    // fix this for call all functions
    expect(dragStartEventSetterSpied).not.toHaveBeenCalled();
    expect(dragMoveEventSetterSpied).not.toHaveBeenCalled();
    expect(dragEndEventSetterSpied).not.toHaveBeenCalled();
    // element.dispatchEvent(new Event('contextmenu'));
  });

  it('take a snapshot of constructor ElementConfig', () => {
    expect.assertions(1);

    expect(ElementConfig).toMatchSnapshot();
  });
});
