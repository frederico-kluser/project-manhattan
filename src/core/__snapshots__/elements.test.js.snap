// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`function ElementBuilder function injectedCssUpdater insert style tag on html head 1`] = `
<style
  type="text/css"
>
  
body {
  background-color: #212121;
}

body {
  margin: 0px;
  padding: 0px;
}

.noselect {
  -khtml-user-select: none; /* Konqueror HTML */
  -moz-user-select: none; /* Old versions of Firefox */
  -ms-user-select: none; /* Internet Explorer/Edge */
  -webkit-touch-callout: none; /* iOS Safari */
  -webkit-user-select: none; /* Safari */
  user-select: none; /* Non-prefixed version, currently supported by Chrome, Edge, Opera and Firefox */
}

.kluser_helper {
  background-color: #212121;
  border-radius: 10px;
  border: 1px solid #FFFFFF;
  height: 30px;
  left: 25px;
  max-width: 30px;
  min-width: 30px;
  opacity: 0.1;
  overflow: hidden;
  position: fixed;
  text-align: center;
  top: 25px;
  transition-timing-function: linear;
  transition: all 0.25s;
  transition: opacity 0.25s, max-width 0.5s;
  width: auto;
}

.kluser_helper:hover {
  opacity: 1;
}

.kluser_helper .material-icons {
  background-color: #212121;
  color: #FFFFFF;
  font-size: 20px;
  left: 0px;
  line-height: 30px;
  position: absolute;
  width: 30px;
  z-index: 1;
}

.kluser_text {
  color: #FFFFFF;
  float: left;
  font-family: 'Roboto Mono', monospace;
  font-helperBubbleDefaultPixelSize: 14px;
  height: 30px;
  line-height: 30px;
  margin-left: 38px;
  padding-right: 16px;
  position: relative;
  text-align: right;
}

.kluser_text:empty {
  display: none;
}

.kluser_input {
  background-color: transparent;
  border: none;
  color: #EEEEEE;
  display: none;
  float: left;
  font-family: 'Roboto Mono', monospace;
  height: 30px;
  line-height: 30px;
  outline: none;
  padding-right: 16px;
  padding: 0px;
}

.kluser_input::placeholder {
  color: #BDBDBD;
}

#mockId1 {
	background-color: #FFFFFF;
	border: 1px solid black;
	height: 100px;
	left: 100px;
	position: relative;
	top: 100px;
	width: 100px;
}

#mockId2 {
	background-color: #FFFFFF;
	border: 1px solid black;
	height: 100px;
	left: 100px;
	position: relative;
	top: 100px;
	width: 100px;
}

#mockId3 {
	background-color: #FFFFFF;
	border: 1px solid black;
	height: 100px;
	left: 100px;
	position: relative;
	top: 100px;
	width: 100px;
}

</style>
`;

exports[`function ElementBuilder take a snapshot of Elements 1`] = `
Object {
  "mockId1": Object {
    "dragEvents": [Function],
    "elementGetter": [Function],
    "modeSetter": [Function],
    "styleGetter": [Function],
    "styleSetter": [Function],
  },
  "mockId2": Object {
    "dragEvents": [Function],
    "elementGetter": [Function],
    "modeSetter": [Function],
    "styleGetter": [Function],
    "styleSetter": [Function],
  },
  "mockId3": Object {
    "dragEvents": [Function],
    "elementGetter": [Function],
    "modeSetter": [Function],
    "styleGetter": [Function],
    "styleSetter": [Function],
  },
}
`;

exports[`function ElementBuilder take a snapshot of constructor ElementConfig 1`] = `[Function]`;

exports[`function ElementBuilder test created elements 1`] = `
<div
  id="mockId3"
/>
`;
